### PR TITLE
yulInterpreter: Add timeout based on the number of interpreted statem…

### PR DESCRIPTION
**Pull request by bshastry**
This PR ensures that we don't flag a bug in the following scenario.

Unoptimized trace times out because of exceeding the step limit

Optimized trace does is valid because the code that exceeded the step limit is redundant and hence optimized out.

Example

```
{
let a,b := foo(calldataload(0),calldataload(32),calldataload(64),calldataload(96),calldataload(128),calldataload(160),calldataload(192),calldataload(224))
sstore(0, a)
sstore(32, b)
function foo(x_0, x_1, x_2, x_3, x_4, x_5, x_6, x_7) -> x_8, x_9
{
for { let i_0 := 0 } lt(i_0, 0x100) { i_0 := add(i_0, 0x20) } {
for { let i_1 := 0 } lt(i_1, 0x100) { i_1 := add(i_1, 0x20) } {
for { let i_2 := 0 } lt(i_2, 0x100) { i_2 := add(i_2, 0x20) } {
for { let i_3 := 0 } lt(i_3, 0x100) { i_3 := add(i_3, 0x20) } {
if sdiv(div(not(1),18446744073709551615),add(1,18446744073709551615)) {}
}
}
}
}
}
}

$ ./fuzzer
Unoptimized trace:
  Interpreter execution step limit reached.

Optimized trace
  SSTORE(0, 0)
  SSTORE(32, 0)
```

In this code, the nested for loops do not have any side effect and are removed by the optimizer. However, the traces of unoptimized code and optimized code differ, resulting in this program being flagged as buggy input. This is clearly a false positive.

### CODE REVIEW

**Review by chriseth — APPROVED** [review]

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp
Comment (christianparpart) [inline test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Instead of hard-coding a string-find on an error message, can we return a status code instead that is reflecting the success and failure type (such as this)? This frees you up from bugs because of foreign string changes and makes the test a little more save (same below).

Comment (bshastry) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Return codes other than zero are reserved by the fuzzer so I can't use them here.

Comment (chriseth) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: But `yulFuzzerUtil::interpret` could return a code - couldn't it?

Comment (bshastry) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Actually, I'm not sure how to use return codes to signal failure in our current architecture. Right now, we throw an exception inside the interpreter when the first of the following happens
  - reached maxSteps (this PR)
  - reached trace size limit
Before throwing the exception, the interpreter appends informative text to the trace for debug purposes.

These exceptions are then caught by `yulFuzzerUtil::interpret` simply relays the trace to the caller i.e., the fuzzing harness. That is the reason I "grep" for it in the harness.

Comment (chriseth) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: You can either check for `m_state.numSteps >= m_state.maxSteps` or throw a different exception in that case. It actually might make sense to throw a different exceptions (where both exceptions share a comment base class) in the "exceptional" termination cases.

**Review by chriseth — DISMISSED** [review]
Please change `yulFuzzerUtil::interpret` to return a status code.

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp
Comment (bshastry) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: > Please change `yulFuzzerUtil::interpret` to return a status code.

Hmm... I guess you mean add a new exception type that shares a comment class (to differentiate between exception classes) and then return a status code on encountering exception class of type "InterpreterStepLimitReached" or something like that.

Comment (chriseth) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Right! Whatever is more convenient :)

Comment (chriseth) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: If you use an exception, you can of course also just catch it outside.

**Comment by bshastry** [comment]
Please review-merge this once #6413 is merged.

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp
Comment (bshastry) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: I used the following approach. Created a flat exception hierarchy that subclasses `dev:Exception` with the following exceptions specific to the yulInterpreter
  - `ExplicitlyTerminated`: triggered during the execution of explicitly terminating opcodes such as `stop`, `revert`, `return` etc.
  - `StepLimitReached`: triggered when the interpreter has evaluated more than `stepLimit` blocks
  - `TraceLimitReached`: triggered when the interpreter has created a trace whose size exceeds the `traceLimit`

Updated exception handling in `yulrun` and the fuzzer code to reflect this change.

**Comment by chriseth** [comment]
Merged #6413

**Comment by chriseth** [comment]
Rebased.

File: test/tools/yulInterpreter/Interpreter.h
Comment (chriseth) [inline test/tools/yulInterpreter/Interpreter.h]: If you keep `InterpreterTerminated` and make it a common base class of the other three, then it is enough to catch `InterpreterTerminated` if you do not care about how it was terminated.

Comment (bshastry) [reply test/tools/yulInterpreter/Interpreter.h]: This is a bit tricky so let me write my notes down:

The fuzz harness does not care if the interpreter
  - was explicitly terminated via stop etc.
  - was terminated because trace limit was exceeded

The assumption behind not caring is that traces should be the same in these cases. The first case (explicit termination) is relatively straightforward. However, I tried to convince myself that the second case also is, but I'm not sure. Essentially, the question is the following: Do traces in the unoptimized and optimized code match exactly at all times? What if there is a side-effect in some dead code that was optimized out?

The test harness does care about
  - termination caused because of exceeding step limit

This is because of the example above.

File: test/libyul/YulInterpreterTest.cpp
Comment (chriseth) [inline test/libyul/YulInterpreterTest.cpp]: If `InterpreterTerminated` would be a common base class of the others, you would only need a single catch statement.

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp
Comment (chriseth) [inline test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Does `ExplicitlyTerminated` lead to a fuzzer failure here (uncaught exception)? Why?

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp
Comment (chriseth) [inline test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Same here.

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp
Comment (chriseth) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Ah, this is a different `interpret` function - I see!

File: test/tools/ossfuzz/yulFuzzerCommon.cpp
Comment (chriseth) [inline test/tools/ossfuzz/yulFuzzerCommon.cpp]: I think it is a bit confusing to catch `ExplicitlyTerminated` here, but leak the other two exceptions.

File: test/tools/yulInterpreter/Interpreter.h
Comment (chriseth) [reply test/tools/yulInterpreter/Interpreter.h]: Please see https://github.com/ethereum/solidity/pull/6398/files#r271915747 for an example of what I meant.

Comment (bshastry) [reply test/tools/yulInterpreter/Interpreter.h]: I think the link is not right, or at least it doesn't point to a code example

File: test/libyul/YulInterpreterTest.cpp
Comment (bshastry) [reply test/libyul/YulInterpreterTest.cpp]: ah, I see. So you are suggesting a hierarchy as follows

```
- InterpreterTerminated
  - ExplicitlyTerminated
  - StepLimitReached
  - TraceLimitReached
```

and then selectively catch exceptions where necessary like the example below (maybe) and catch the base class where selection is not needed e.g., tests.

```
try {
  ...
}
catch (InterpreterTerminated const& it)
{
  if (dynamic_cast<StepLimitReached>(it))
    return 0;
}
```

File: test/tools/ossfuzz/yulFuzzerCommon.cpp
Comment (bshastry) [reply test/tools/ossfuzz/yulFuzzerCommon.cpp]: I agree. I can propagate all catches to the fuzzing harness itself for clarity.

File: test/libyul/YulInterpreterTest.cpp
Comment (chriseth) [reply test/libyul/YulInterpreterTest.cpp]: Exactly! Your example above could also be:
```
try
{
...
}
catch (StepLimitReached const& _e)
{
...
}
catch (InterpreterTerminated const& _e)
{
... // all other cases
}
```
Perhaps the type should be named something like `InterpreterTerminatedGeneric` to make it more obvious that it is a generic base class.

Comment (bshastry) [reply test/libyul/YulInterpreterTest.cpp]: done.

File: test/tools/yulInterpreter/Interpreter.h
Comment (chriseth) [inline test/tools/yulInterpreter/Interpreter.h]:
```suggestion
class InterpreterTerminatedGeneric: public dev::Exception
```
(also for the others)

**Review by chriseth — CHANGES_REQUESTED** [review]
Fine apart from the build failure.

**Review by chriseth — APPROVED** [review]

**Pull request merged**
